agora/Agora/Stake/Scripts.hs

```diff
@@ -378,10 +383,36 @@ stakeValidator stake =
                             PStakeDatum
                             ( #stakedAmount .= stakeDatum.stakedAmount
                                 .& #owner .= stakeDatum.owner
+                                .& #delegatedTo .= stakeDatum.delegatedTo
                                 .& #lockedBy .= pfield @"lockedBy" # stakeOut
                             )
                      in stakeOut #== templateStakeDatum
 
+                setDelegate <- pletC $
+                  plam $ \maybePkh -> unTermCont $ do
+                    pguardC
+                      "Owner signs this transaction"
+                      ownerSignsTransaction
+
+                    pguardC "A UTXO must exist with the correct output" $
+                      let correctOutputDatum =
+                            stakeOut
+                              #== mkRecordConstr
+                                PStakeDatum
+                                ( #stakedAmount .= stakeDatum.stakedAmount
+                                    .& #owner .= stakeDatum.owner
+                                    .& #delegatedTo .= pdata maybePkh
+                                    .& #lockedBy .= stakeDatum.lockedBy
+                                )
+                          valueCorrect = ownOutputValueUnchanged
+                       in foldl1
+                            (#&&)
+                            [ ptraceIfFalse "valueCorrect" valueCorrect
+                            , ptraceIfFalse "datumCorrect" correctOutputDatum
+                            ]
+
+                    pure $ popaque (pconstant ())
+
                 pure $
                   pmatch stakeRedeemer $ \case
                     PRetractVotes _ -> unTermCont $ do
```

Comment:
Definitely add `delegateSignsTransaction` here too. Imagine if the delegatee makes a mistake, or something very important happens. This reminds me that we likely want to support generalized authority here too :dizzy_face:. That's for another PR though :smile: